File: src/main/java/br/com/dusty/dkits/listener/gameplay/PlayerDeathListener.java
package br.com.dusty.dkits.listener.gameplay;

import br.com.dusty.dkits.gamer.Gamer;
import br.com.dusty.dkits.util.TaskUtils;
import br.com.dusty.dkits.util.protocol.ProtocolUtils;
import org.bukkit.entity.Player;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.PlayerDeathEvent;

import java.lang.reflect.Field;

public class PlayerDeathListener implements Listener {
	
	private static Class class_PacketPlayInClientCommand;
	private static Field field_PacketPlayInClientCommand_a;
	
	private static Object[] enum_EnumClientCommand_values;
	
	static {
		try{
			class_PacketPlayInClientCommand = Class.forName(ProtocolUtils.NMS_PACKAGE + ProtocolUtils.NMS_VERSION + ".PacketPlayInClientCommand");
			field_PacketPlayInClientCommand_a = ProtocolUtils.getAccessibleField(class_PacketPlayInClientCommand, "a");
			
			Class enum_EnumClientCommand = Class.forName(ProtocolUtils.NMS_PACKAGE + ProtocolUtils.NMS_VERSION + ".PacketPlayInClientCommand$EnumClientCommand");
			enum_EnumClientCommand_values = enum_EnumClientCommand.getEnumConstants();
		}catch(ClassNotFoundException | NoSuchFieldException e){
			e.printStackTrace();
		}
	}
	
	public void onPlayerDeath(PlayerDeathEvent event) {
		Player player = event.getEntity();
		
		Gamer gamer = Gamer.of(player);
		
		try{
			final Object object_PacketPlayInClientCommand = class_PacketPlayInClientCommand.newInstance();
			field_PacketPlayInClientCommand_a.set(object_PacketPlayInClientCommand, enum_EnumClientCommand_values[0]);
			
			TaskUtils.sync(() -> {
				ProtocolUtils.sendPacket(object_PacketPlayInClientCommand, player);
				
				//TODO: If player was on MiniHG, send to Lobby.
				gamer.sendToWarp(gamer.getWarp());
			});
		}catch(InstantiationException | IllegalAccessException e){
			e.printStackTrace();
		}
	}
}

File: src/main/kotlin/br/com/dusty/dkits/listener/gameplay/PlayerDeathListener.kt
package br.com.dusty.dkits.listener.gameplay

import br.com.dusty.dkits.util.Tasks
import br.com.dusty.dkits.util.gamer.gamer
import br.com.dusty.dkits.warp.Warp
import br.com.dusty.dkits.warp.Warps
import org.bukkit.Material.*
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.PlayerDeathEvent

object PlayerDeathListener: Listener {

	val ALLOWED_DROPS = arrayOf(RED_MUSHROOM, BROWN_MUSHROOM, BOWL, MUSHROOM_SOUP, WOOD_SWORD, STONE_SWORD)

	@EventHandler
	fun onPlayerDeath(event: PlayerDeathEvent) {
		val player = event.entity
		val gamer = player.gamer()

		if (gamer.warp.overrides(event)) return

		event.deathMessage = null
		event.drops.removeIf { it.type !in ALLOWED_DROPS }

		if (gamer.isCombatTagged()) {
			val combatPartner = gamer.combatPartner!!

			if (gamer.warp.type != Warp.EnumWarpType.EVENT) combatPartner.kill(gamer)

			if (combatPartner.warp.durabilityBehavior == Warp.EnumDurabilityBehavior.REGEN_ON_KILL) combatPartner.player.inventory.apply {
				itemInHand?.durability = 0
				armorContents.forEach { it?.durability = 0 }
			}
		}

		gamer.resetKillStreak()

		Tasks.sync(Runnable {
			player.spigot().respawn()

			gamer.sendToWarp(if (gamer.warp.type == Warp.EnumWarpType.EVENT) Warps.LOBBY else gamer.warp, true, false)
		})
	}
}
